cli/replicaset/cconfig.go

```diff
@@ -204,3 +218,408 @@ func updateCConfigInstances(replicaset *Replicaset, topology cconfigTopology) {
 		}
 	}
 }
+
+// Promote promotes an instance in the application.
+func (c *CConfigApplication) Promote(ctx PromoteCtx) error {
+	replicasets, err := c.GetReplicasets()
+	if err != nil {
+		return fmt.Errorf("failed to get replicasets: %w", err)
+	}
+
+	var (
+		targetReplicaset Replicaset
+		targetInstance   Instance
+		found            bool
+	)
+loop:
+	for _, replicaset := range replicasets.Replicasets {
+		for _, instance := range replicaset.Instances {
+			if instance.Alias == ctx.InstName {
+				targetReplicaset = replicaset
+				targetInstance = instance
+				found = true
+				break loop
+			}
+		}
+	}
+	if !found {
+		return fmt.Errorf("instance %q not found in a configured replicaset", ctx.InstName)
+	}
+	if !targetInstance.InstanceCtxFound {
+		return fmt.Errorf("instance %q should be online", ctx.InstName)
+	}
+
+	var instances []running.InstanceCtx
+	var unfound []string
+	for _, inst := range targetReplicaset.Instances {
+		if !inst.InstanceCtxFound {
+			unfound = append(unfound, inst.Alias)
+		} else {
+			instances = append(instances, inst.InstanceCtx)
+		}
+	}
+	if len(unfound) > 0 && !ctx.SkipUnfoundInstances {
+		return fmt.Errorf("all instances in the target replicaset should "+
+			"be online, could not connect to: %s", strings.Join(unfound, ","))
+	}
+
+	isConfigPatched, err := cconfigPromote(targetInstance.InstanceCtx, ctx)
+	if err != nil {
+		return err
+	}
+	if isConfigPatched {
+		err = reloadCConfig(instances)
+	}
+	return err
+}
+
+// cconfigPromoteElection tries to promote an instance via `box.ctl.promote()`.
+func cconfigPromoteElection(evaler connector.Evaler, timeout int) error {
+	args := []any{timeout}
+	opts := connector.RequestOpts{}
+	_, err := evaler.Eval(cconfigPromoteElectionBody, args, opts)
+	return err
+}
+
+// CConfigPublisher publishes the config.
+type CConfigPublisher interface {
+	// Publish publishes the config into a specific key that contains the config prefix.
+	Publish(key string, data []byte, revision int64) error
+}
+
+// CConfigSource is a cluster config source.
+type CConfigSource struct {
+	dataCollector libcluster.DataCollector
+	publisher     CConfigPublisher
+}
+
+// NewConfigSource creates CConfigSource.
+func NewCConfigSource(dataCollector libcluster.DataCollector,
+	publisher CConfigPublisher) CConfigSource {
+	return CConfigSource{
+		dataCollector: dataCollector,
+		publisher:     publisher,
+	}
+}
```

Comment:
It looks like a different case in general. I think it's better to move it in a separate file, `source.go` or `configsource.go` as example.

Reply:
Moved.